packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -110,7 +138,43 @@ class ProtocolDiscovery {
     if (ipv6) {
       hostUri = hostUri.replace(host: InternetAddress.loopbackIPv6.host);
     }
-
     return hostUri;
   }
 }
+
+/// This transformer will produce an event at most once every [waitDuration].
+///
+/// For example, consider a `waitDuration` of `10ms`, and list of event names
+/// and arrival times: `a (0ms), b (5ms), c (11ms), d (21ms)`.
+/// The events `c` and `d` will be produced as a result.
+StreamTransformer<S, S> _throttle<S>({
+  @required Duration waitDuration,
+}) {
+  assert(waitDuration != null);
+
+  S latestLine;
+  int lastExecution;
+  Future<void> throttleFuture;
+
+  return StreamTransformer<S, S>
+    .fromHandlers(
+      handleData: (S value, EventSink<S> sink) {
+        latestLine = value;
+
+        final int currentTime = DateTime.now().millisecondsSinceEpoch;
+        lastExecution ??= currentTime;
```

Comment:
Should `lastExecution` be initialized earlier? The way this is written it looks like the first event will always be emitted no earlier than `throttleDuration` ms after the first event comes in. Is that what you had in mind?

Reply:
I tried it, but I got very different results since the initial value is time sensitive. The time spent between the initialization and the first event was in some cases greater than the actual wait time.